### DISCUSSION TITLE
gpsd-compatible output possible?

### DISCUSSION THREAD

**Discussion by juhench** [strukturart/o.map#54] [Ideas]
Hi! I've been using (and liking) o.map since I switched to KaiOS which I use in combination with a GPD micropc as a smartphone replacement. Is there a technical way to run a web server on the KaiOS phone that would allow polling gpsd-compatible content (current coordinate, metadata) information from the PC? The phone would then be in hotspot mode.

I am not sure whether this is at all possible with KaiOS privileged apps. If possible, this would make KaiOS devices cost-effective GPS receivers with a long battery life. If it is technically possible, this would be a convenient feature to add to o.map.

If it is impossible to open a port on the KaiOS device, this could be realized by sending HTTP queries to the PC and running a small service on the PC (e.g. cherrypy) that processes these queries. The queries could then be internally forwarded to gpsd on the PC to enable compatibility with existing applications.

For a start, a destination IP could be set in o.map to receive the queries at a definable poll rate.

**Comment by strukturart** [comment]
what about using webRTC to trnasfer the data ?

**Reply by juhench** [reply]
webRTC is a great idea. Would it be possible to pass over a set of data that would be then translated to the parameters that [gpsd](https://gpsd.gitlab.io/gpsd/gpsd.html) specifies? To ensure compatibility with let's say foxtrotgps one could fake potentially missing values.

Have you tried webRTC before on KaiOS, is the "Firefox" on the phone able to deal with it?  That would mean a small standalone app should do the job, isn't it?

**Reply by strukturart** [reply]
I wrote a webRTC app for KaiOS:
https://github.com/strukturart/flop

It works well with KaiOS 3.x, but with KaiOS 2.x it is sometimes difficult to establish a connection.

If you have the possibility to sideload an app, I can provide you with a version if you want to test it.

**Reply by juhench** [reply]
Thanks for sharing. I have an E242S, i.e., I would try installing "flop" via the developer portal. The E242S does not have a complete adb binary, unfortunately. I will try testing with the code you have in your "flop" repo and maybe bother you again if I don't get that to work.

Thanks again for your help.

**Comment by strukturart** [comment]
I can add you as a tester if you send me your IMEI via email.

**Reply by juhench** [reply]
even better! Will do...

**Reply by juhench** [reply]
sent to print@... as in your GitHub Account. Thanks!

**Reply by strukturart** [reply]
done

**Reply by juhench** [reply]
great, thanks. Will try!

**Reply by juhench** [reply]
Which category did you use? I usually use "shopping". Search 'flop' or other development apps does not work on my E242S.

**Reply by strukturart** [reply]
category: social

**Comment by strukturart** [comment]
You can establish the connection between KaiOS devices via QR code, which works well, but less well between two KaiOS 2.x devices.

this website can also run locally on your computer, which you can then use to exchange data: 
https://flop.bhackers.uber.space/#!/

**Reply by juhench** [reply]
Looking forward to try - looks like a good starting point for me. Will give feedback!

**Reply by juhench** [reply]
The E242S and a laptop got connected immediately, however, the connection is not very stable, as already stated by you. This is a very good starting point for me, thanks again!

**Comment by strukturart** [comment]
With flop it would be possible to display the current position of a device live on a map. To do this I would have to integrate leaflet into flop and use the GPS interval as a trigger to update the marker on the map. You could then read out the marker data and process it further.

The basic requirement would of course be that flop can establish a more stable connection, possibly because the KaiOS device goes into sleep mode.
